File: ios/Memorezar/Data/Services/LanguageService.swift
import Foundation
import SwiftUI

/// Fetches language display config (colors) from Supabase.
/// Falls back to hardcoded defaults if offline.
final class LanguageService: ObservableObject {
    static let shared = LanguageService()

    @Published private(set) var languages: [String: LanguageConfig] = LanguageService.defaults

    private var languagesURL: URL {
        URL(string: "\(SupabaseConfig.projectURL)/rest/v1/languages?select=*")!
    }

    struct LanguageConfig {
        let displayName: String
        let color: Color
        let textColor: Color
    }

    private static let defaults: [String: LanguageConfig] = [
        "en": LanguageConfig(displayName: "English", color: .blue, textColor: .white),
        "es": LanguageConfig(displayName: "Spanish", color: .yellow, textColor: .black),
        "fr": LanguageConfig(displayName: "French", color: .red, textColor: .white),
        "it": LanguageConfig(displayName: "Italian", color: .green, textColor: .black),
        "de": LanguageConfig(displayName: "German", color: .orange, textColor: .black),
        "pt": LanguageConfig(displayName: "Portuguese", color: .green, textColor: .black),
        "ar": LanguageConfig(displayName: "Arabic", color: .green, textColor: .white),
    ]

    func fetchLanguages() async {
        var request = URLRequest(url: languagesURL)
        request.timeoutInterval = 10
        for (key, value) in SupabaseConfig.headers {
            request.setValue(value, forHTTPHeaderField: key)
        }

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else { return }

            let rows = try JSONDecoder().decode([RemoteLanguageRow].self, from: data)
            var built: [String: LanguageConfig] = [:]
            for row in rows {
                built[row.code] = LanguageConfig(
                    displayName: row.display_name,
                    color: Color(hex: row.color),
                    textColor: Color(hex: row.text_color)
                )
            }
            let finalLanguages = built
            await MainActor.run {
                self.languages = finalLanguages
            }
        } catch {
            print("[LanguageService] Fetch failed: \(error.localizedDescription)")
        }
    }

    func color(for code: String) -> Color {
        languages[code]?.color ?? .indigo
    }

    func textColor(for code: String) -> Color {
        languages[code]?.textColor ?? .white
    }
}

private struct RemoteLanguageRow: Codable {
    let code: String
    let display_name: String
    let color: String
    let text_color: String
}

extension Color {
    init(hex: String) {
        let hex = hex.trimmingCharacters(in: CharacterSet(charactersIn: "#"))
        var int: UInt64 = 0
        Scanner(string: hex).scanHexInt64(&int)
        let r = Double((int >> 16) & 0xFF) / 255
        let g = Double((int >> 8) & 0xFF) / 255
        let b = Double(int & 0xFF) / 255
        self.init(red: r, green: g, blue: b)
    }
}

File: memorezar-android/app/src/main/java/com/memorezar/app/data/services/LanguageService.kt
package com.memorezar.app.data.services

import android.util.Log
import androidx.compose.runtime.compositionLocalOf
import androidx.compose.ui.graphics.Color
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withContext
import org.json.JSONArray
import java.net.HttpURLConnection
import java.net.URL

/**
 * Mirrors iOS's LanguageService. Fetches per-language colors from the Supabase
 * `languages` table at app start, falling back to hardcoded defaults if offline
 * or before the first fetch returns.
 *
 * Kept as a kotlin `object` (no DI) so the existing `languageColor()` helper in
 * RecitationScreen can read the cached map without plumbing a dependency
 * through every composable.
 */
object LanguageService {

    data class Config(
        val displayName: String,
        val color: Color,
        val textColor: Color,
    )

    private val defaults: Map<String, Config> = mapOf(
        "en" to Config("English", Color(0xFF2196F3), Color.White),
        "es" to Config("Spanish", Color(0xFFFFC107), Color.Black),
        "fr" to Config("French", Color(0xFFF44336), Color.White),
        "it" to Config("Italian", Color(0xFF34C759), Color.Black),
        "de" to Config("German", Color(0xFFFF9800), Color.Black),
        "pt" to Config("Portuguese", Color(0xFF34C759), Color.Black),
        "ar" to Config("Arabic", Color(0xFF34C759), Color.White),
    )

    private val _languages = MutableStateFlow(defaults)
    val languages: StateFlow<Map<String, Config>> = _languages.asStateFlow()

    fun color(code: String): Color =
        _languages.value[code.lowercase()]?.color ?: Color(0xFF7A71F0)

    fun textColor(code: String): Color =
        _languages.value[code.lowercase()]?.textColor ?: Color.White

    /** One-shot fetch; called from HomeScreen on app launch. Safe to call repeatedly. */
    suspend fun fetchLanguages() = withContext(Dispatchers.IO) {
        try {
            val url = URL("${SupabaseConfig.PROJECT_URL}/rest/v1/languages?select=*")
            val conn = (url.openConnection() as HttpURLConnection).apply {
                setRequestProperty("apikey", SupabaseConfig.ANON_KEY)
                setRequestProperty("Authorization", "Bearer ${SupabaseConfig.ANON_KEY}")
                setRequestProperty("Accept", "application/json")
                connectTimeout = 10_000
                readTimeout = 10_000
            }
            if (conn.responseCode != 200) {
                Log.w("LanguageService", "fetch returned ${conn.responseCode}")
                return@withContext
            }
            val body = conn.inputStream.bufferedReader().use { it.readText() }
            val rows = JSONArray(body)
            val built = mutableMapOf<String, Config>()
            for (i in 0 until rows.length()) {
                val row = rows.getJSONObject(i)
                val code = row.optString("code").lowercase()
                if (code.isEmpty()) continue
                built[code] = Config(
                    displayName = row.optString("display_name", code),
                    color = parseHex(row.optString("color")) ?: defaults[code]?.color ?: Color(0xFF7A71F0),
                    textColor = parseHex(row.optString("text_color")) ?: defaults[code]?.textColor ?: Color.White,
                )
            }
            if (built.isNotEmpty()) {
                _languages.value = built
            }
        } catch (e: Exception) {
            Log.w("LanguageService", "fetch failed: ${e.message}")
        }
    }

    private fun parseHex(hex: String?): Color? {
        if (hex.isNullOrBlank()) return null
        val cleaned = hex.trim().removePrefix("#")
        return try {
            val v = cleaned.toLong(16)
            // Color(Int) takes packed 0xAARRGGBB. Add full alpha for 6-digit hex.
            val argb = if (cleaned.length == 6) (v or 0xFF000000L) else v
            Color(argb.toInt())
        } catch (_: NumberFormatException) {
            null
        }
    }
}

/** Composition-local hook so the Compose `languageColor`/`languageTextColor`
 *  helpers can subscribe to the StateFlow and recompose when the fetch arrives. */
val LocalLanguages = compositionLocalOf { LanguageService.languages.value }
